sdk/core/core/src/az_http_response.c

```diff
@@ -268,3 +268,36 @@ void _az_http_response_reset(az_http_response* http_response)
   az_result result = az_http_response_init(http_response, http_response->_internal.http_response);
   (void)result;
 }
+
+// internal function to get az_http_response remainder
+static az_span _az_http_response_get_remaining(az_http_response const* response)
+{
+  return az_span_slice_to_end(response->_internal.http_response, response->_internal.written);
+}
+
+AZ_NODISCARD az_result az_http_response_write_span(az_http_response* response, az_span write)
+{
+  AZ_PRECONDITION_NOT_NULL(response);
+
+  az_span remaining = _az_http_response_get_remaining(response);
+  int32_t write_size = az_span_size(write);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining, write_size);
+
+  remaining = az_span_copy(remaining, write);
+  response->_internal.written += write_size;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_http_response_write_u8(az_http_response* response, uint8_t byte)
+{
+  AZ_PRECONDITION_NOT_NULL(response);
+
+  az_span remaining = _az_http_response_get_remaining(response);
+  AZ_RETURN_IF_NOT_ENOUGH_SIZE(remaining, 1);
```

Comment:
Is it feasible to make this a precondition instead? I don't think so, but asking to verify.

Reply:
don't think so.
This function is usually used by adding a zero at the end of a response. We can't be sure about the size of the response will never be such that compiling away the precondition we remain safe

Reply:
> We can't be sure about the size of the response will never be such that compiling away the precondition we remain safe

- Can the caller check that the response is large enough up front, relatively easily? I think, yes. The caller should be checking the size, anyway, right? Is it easy for them to do so?
- If we return not enough size, what should the caller do with that error condition? Is the error meaningful to the caller? I am not sure.

Whether we keep the method "safe" when preconditions are off is an implementation detail and orthogonal to whether our API shape should return az_result and if the contract states the response buffer must be large enough.

Reply:
>Can the caller check that the response is large enough up front, relatively easily? I think, yes. The caller should be checking the size, anyway, right? Is it easy for them to do so?

I don't think so.. How can the caller do this?  This depends on the customer subscription and data.  
We might be able to see content-length if the service is returning this header, but if not, we just keep pulling from wire until we either get all response of failed and return overflow. When overflow, customer needs to implement the memory re-allocation for a bigger buffer and try the request again.

>Whether we keep the method "safe" when preconditions are off is an implementation detail and orthogonal to whether our API shape should return az_result and if the contract states the response buffer must be large enough.

My understanding is that release products will always be with no preconditions. So it is on developing and testing when preconditions will help us to find bugs and fix them before release. For some cases, we can be sure that testing can cover all scenarios (like ranges), but for others, we can't

Reply:
Discussed offline. We don't want to make it a precondition since that would rely on exposing a concept that is currently internal. 

Marking as resolved.